src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -2168,13 +2218,35 @@ public int GetLowerBound(int dimension)
 
             if (this.Rank != 1)
                 throw new PlatformNotSupportedException(SR.Rank_MultiDimNotSupported);
+#endif
             if (dimension != 0)
                 throw new IndexOutOfRangeException();
             return 0;
         }
 
         public int GetUpperBound(int dimension)
         {
+#if REAL_MULTIDIM_ARRAYS
+            if (!this.EETypePtr.IsSzArray)
+            {
+                if ((dimension >= Rank) || (dimension < 0))
+                    throw new IndexOutOfRangeException();
+
+                unsafe
+                {
+                    fixed (int* pNumComponents = &_numComponents)
+                    {
+                        // Upper bounds follow after _numComponents.
+#if BIT64
+                        // _numComponents is padded on 64-bit systems.
+                        return *(pNumComponents + 1 + 1 + dimension) - 1;
```

Comment:
Should it match what CoreCLR does?

https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/classlibnative/bcltype/arraynative.cpp#L72